src/v/datalake/translation/partition_translator.cc

```diff
@@ -0,0 +1,273 @@
+
+#include "datalake/translation/partition_translator.h"
+
+#include "cluster/partition.h"
+#include "datalake/batching_parquet_writer.h"
+#include "datalake/coordinator/frontend.h"
+#include "datalake/coordinator/translated_offset_range.h"
+#include "datalake/data_writer_interface.h"
+#include "datalake/logger.h"
+#include "datalake/record_multiplexer.h"
+#include "datalake/translation/state_machine.h"
+#include "kafka/server/partition_proxy.h"
+#include "kafka/utils/txn_reader.h"
+#include "resource_mgmt/io_priority.h"
+#include "ssx/future-util.h"
+
+namespace datalake::translation {
+
+static constexpr std::chrono::milliseconds translation_jitter{500};
+static constexpr int max_attempts = 5;
+constexpr ::model::timeout_clock::duration wait_timeout = 5s;
+constexpr size_t max_rows_per_row_group = std::numeric_limits<size_t>::max();
+constexpr size_t max_bytes_per_row_group = std::numeric_limits<size_t>::max();
+
+partition_translator::partition_translator(
+  ss::lw_shared_ptr<cluster::partition> partition,
+  ss::sharded<coordinator::frontend>* frontend,
+  ss::sharded<features::feature_table>* features,
+  std::chrono::milliseconds translation_interval,
+  ss::scheduling_group sg,
+  size_t reader_max_bytes,
+  ssx::semaphore& parallel_translations)
+  : _term(partition->raft()->term())
+  , _partition(std::move(partition))
+  , _frontend(frontend)
+  , _features(features)
+  , _jitter{translation_interval, translation_jitter}
+  , _max_bytes_per_reader(reader_max_bytes)
+  , _parallel_translations(parallel_translations)
+  , _writer_scratch_space(std::filesystem::temp_directory_path())
+  , _logger(prefix_logger{
+      datalake_log, fmt::format("{}-term-{}", _partition->ntp(), _term)}) {
+    _stm = _partition->raft()
+             ->stm_manager()
+             ->get<datalake::translation::translation_stm>();
+    vassert(_stm, "No translation stm found for {}", _partition->ntp());
+    ssx::repeat_until_gate_closed(_gate, [this, sg] {
+        return ss::with_scheduling_group(sg, [this] { return do_translate(); });
+    });
+    vlog(_logger.debug, "started partition translator in term {}", _term);
+}
+
+std::chrono::milliseconds partition_translator::translation_interval() const {
+    return _jitter.base_duration();
+}
+
+void partition_translator::reset_translation_interval(
+  std::chrono::milliseconds new_base) {
+    _jitter = jitter_t{new_base, translation_jitter};
+    vlog(
+      _logger.info,
+      "Iceberg translation interval reset to: {}",
+      _jitter.base_duration());
+}
+
+ss::future<> partition_translator::stop() {
+    vlog(_logger.debug, "stopping partition translator in term {}", _term);
+    auto f = _gate.close();
+    _as.request_abort();
+    co_await std::move(f);
+}
+
+kafka::offset partition_translator::min_offset_for_translation() const {
+    auto min_offset = _partition->raft_start_offset();
+    return model::offset_cast(
+      _partition->get_offset_translator_state()->from_log_offset(min_offset));
```

Comment:
There are a couple of palaces we have to remember to translate state - the different type helps a little here, but should we just keep a partition_proxy instead of a raw cluster::partition member in this class? IIRC that will handle all the translation, which it seems good to try and keep that in a single layer.